mustache.go

```diff
@@ -31,19 +31,31 @@ func (es ErrorSlice) Error() string {
 	return b.String()
 }
 
+func maybeAddNode(seenNodes map[node]struct{}, newNode node) bool {
+	if _, ok := seenNodes[newNode]; !ok {
+		seenNodes[newNode] = struct{}{}
+		return true
+	}
+	return false
+}
+
+func nodeAlreadyVisitedError(node string) error {
+	return fmt.Errorf("the node %s has been already visited in the mustache node graph", node)
+}
+
 // The node type is the base type that represents a node in the parse tree.
 type node interface {
 	// The render function should be defined by any type wishing to satisfy the
 	// node interface. Implementations should be able to render itself to the
 	// w Writer with c given as context.
-	render(t *Template, w *writer, c ...interface{}) error
+	render(t *Template, w *writer, seenNodes map[node]struct{}, c ...interface{}) error
 }
```

Comment:
Shouldn't we also add a String() method? (so that all implementations are forced to implement it?)

Reply:
Sure I don't mind adding it 😄